cpp/src/umap/runner.cuh

```diff
@@ -263,22 +358,30 @@ void _fit_supervised(const raft::handle_t& handle,
                      const umap_inputs& inputs,
                      UMAPParams* params,
                      value_t* embeddings,
-                     raft::sparse::COO<float, int>* graph)
+                     raft::host_coo_matrix<float, int, int, uint64_t>& host_graph)
 {
   raft::common::nvtx::range fun_scope("umap::supervised::fit");
 
-  cudaStream_t stream = handle.get_stream();
+  auto stream = handle.get_stream();
   ML::default_logger().set_level(params->verbosity);
 
-  UMAPAlgo::_get_graph_supervised<value_idx, value_t, umap_inputs, nnz_t, TPB_X>(
-    handle, inputs, params, graph);
+  raft::sparse::COO<value_t> trimmed_graph(stream);
+  {
+    raft::sparse::COO<value_t> graph(stream);
+    UMAPAlgo::_get_graph_supervised<value_idx, value_t, umap_inputs, nnz_t, TPB_X>(
+      handle, inputs, params, &graph);
+
+    copy_device_graph_to_host(handle, graph, host_graph);
+
+    trim_graph(handle, graph, trimmed_graph, inputs.n, params->n_epochs);
+  }
```

Comment:
I assume the scope here is to drop `graph` once the trimming is done to reduce device memory pressure? If so, can you add a small comment noting that?

Also, is there any way the trimming could be done in place? Last time I checked this was one of the memory highpoints, so reducing duplications of the graph here would be beneficial if easy.

Reply:
> Performing the operation on host might be relatively fast and would avoid doubling VRAM use here, but it would be probably much slower than doing it on device with a copy.

Generally, host copies are super quick and will never be the bottleneck. We have seen this in other uses like NN Descent (cc @jinsolp).

@jcrist does this section of the code spike higher than peak memory?